python_modules/dagster/dagster/_config/structured_config/typing_utils.py

```diff
@@ -0,0 +1,49 @@
+from typing import TYPE_CHECKING, Any, Optional, TypeVar, Union, cast
+
+if TYPE_CHECKING:
+    from dagster._config.structured_config import PartialResource
+
+Self = TypeVar("Self", bound="AllowPartialResourceInitParams")
+
+
+class AllowPartialResourceInitParams:
+    """
+    Implementation of the Python descriptor protocol (https://docs.python.org/3/howto/descriptor.html)
+    to adjust the types of resource inputs and outputs, e.g. resource dependencies can be passed in
+    as PartialResources or Resources, but will always be returned as Resources.
+
+    For example, given a resource with the following signature:
+
+    .. code-block:: python
+
+        class FooResource(Resource):
+            bar: BarResource
+
+    The following code will work:
+
+    .. code-block:: python
+
+        # Types as PartialResource[BarResource]
+        partial_bar = BarResource.configure_at_runtime()
+
+        # bar parameter takes BarResource | PartialResource[BarResource]
+        foo = FooResource(bar=partial_bar)
+
```

Comment:
One point of explanation here is that something has to happen between construction and the print state for `foo.bar` to return a `BarResource`. Some code has to go through and replace the partial with the full thing.